### PR TITLE
Send users to new CSP 20-21 URLs for lockable lessons with lesson plans and redirect old URL

### PR DESCRIPTION
**Pull request by dmcavoy**
This start sending users to new urls for CSP 20-21 special lessons `/s/csp1-2020/stage/14/puzzle/1/page/1` for example. It also redirects the old urls to the new url. 

This is part **2** of 5 for moving CSP 20-21 lockable lessons with lesson plans over to new URLs at `/stage` instead of `/lockable`. The steps are:

1. (https://github.com/code-dot-org/code-dot-org/pull/39035) Get script_level by absolute position instead of relative position for CSP 20-21 lockable lessons with lesson plans
2. (**THIS PR**) Start using `/stage` URLs for CSP 20-21 lockable lesson with lesson plans and redirect old `/lockable` URL to the new `/stage` URL.
3. (https://github.com/code-dot-org/code-dot-org/pull/39129) Set has_lesson_plan and update name on CSP 20-21 lockable lessons with lesson plans. This will update relative_position as well
4. (https://github.com/code-dot-org/code-dot-org/pull/39130) Add the permanent redirects to routes.rb for CSP 20-21 lockable lessons with lesson plans
5. (https://github.com/code-dot-org/code-dot-org/pull/39131) Remove the controller code that creates the redirects for CSP 20-21 lockable lessons with lesson plans

## Links

- [spec](https://docs.google.com/document/d/1sAAVnwv3rXQEK-qa8JnkNXFfZ0vjWQyzJ9Yx_pG4bdg/edit#heading=h.my8ycsh5b29w)
- [jira](https://codedotorg.atlassian.net/browse/PLAT-735)

## Testing story

Added two integration tests. One for csp1-2020 and one for csp2-2020 which cover the two different cases we have to look for. CSP 1 has 2 lockable lessons where as all the other units have 1 lockable lesson.

### CODE REVIEW

File: dashboard/test/integration/scripts_test.rb
Comment (davidsbailey) [inline dashboard/test/integration/scripts_test.rb:24]: there is a bit too much hidden here, e.g. is this the old url, or the new one? I think it would be better to hit the url explicitly, otherwise the url this test is covering will change as you make various code changes.

File: dashboard/test/integration/scripts_test.rb
Comment (davidsbailey) [inline dashboard/test/integration/scripts_test.rb:3]: nice move adding integration tests! the ability to specify the path and follow the redirects seems like a big win. I would suggest focusing your efforts here, since this lets you really test that things are working end to end.

**Review by davidsbailey — COMMENTED** [review]
Hi Dani, a few top level comments:

1. testing this feature thoroughly at each step will be very important.
2. due to the short lived nature of each of these changes, I think you have some leeway in terms of how much you do in an automated fashion vs manual. 
3. the controller tests have a shortcoming, which is that they don't (can't) verify that the right level is loaded after following a redirect. the integration tests do not have this limitation
4. there are so many cases that it is hard to figure out which ones are fully covered by automated testing and which are not.

I've created a gsheet with 8 urls that seem to me like they cover all of the codepaths I am worried about: https://docs.google.com/spreadsheets/d/1J47uKxCGfK4SUdq2qeiOM8HAHIUgsKvnlhCy_pLxgEg/edit#gid=0

I see two paths forward:
1. commit to a full manual test pass of all 8 urls on each PR before it is merged. you can keep the tests you have, with the understanding that they do not provide complete coverage.
2. add integration tests which check all 8 urls, following redirects as necessary, and verifying that the correct level page ultimately gets loaded. if you go this route, you can probably remove the controller tests.

because there are only levelgroups in the csp assessment units, it does not seem important to test the external level case.

Feel free to get other opinions here... there's a chance I am the most worried person in the room about what we are trying to do here. Apologies for not seeing the shortcomings of the controller tests earlier.

Let's discuss before you move forward.

File: dashboard/test/integration/scripts_test.rb
Comment (davidsbailey) [reply dashboard/test/integration/scripts_test.rb:24]: after offline conversation... if these tests are just serving as backup for thorough manual testing, then it seems ok to test this, and this will likely have less churn than using an explicit url. OTOH if you are trying to get full test coverage via integration tests, then I would test each literal url explicitly.

**Review by davidsbailey — APPROVED** [review]
LGTM assuming a comprehensive manual test pass. :-)

**Comment by dmcavoy** [comment]
Manual Testing Report

| URL | Tested? |
| -- | -- |
|/s/csp1-2020/stage/14/puzzle/1| X |
|/s/csp1-2020/stage/14/puzzle/1/page/1| X |
|/s/csp1-2020/lockable/2/puzzle/1| X |
|/s/csp1-2020/lockable/2/puzzle/1/page/1| X |
|/s/csp4-2020/stage/15/puzzle/1| X |
|/s/csp4-2020/stage/15/puzzle/2| X |
|/s/csp4-2020/stage/15/puzzle/2/page/1| X |
|/s/csp4-2020/lockable/1/puzzle/1| X |
|/s/csp4-2020/lockable/1/puzzle/2| X |
|/s/csp4-2020/lockable/1/puzzle/2/page/1| X |

**Review by bethanyaconnor — APPROVED** [review]
This change looks good to me but I'm struggling to find where the redirect is happening. Am I missing something?

**Comment by dmcavoy** [comment]
> This change looks good to me but I'm struggling to find where the redirect is happening. Am I missing something?

It's definitely tricky. There is a bunch of logic in the show action in the script levels controller that uses build_script_level_path. What ends up happening is that it checks if they URL that came in is the same as the URL that build_script_level_path would make for the script level. If not it will use the build_script_level_path url instead

**Comment by bethanyaconnor** [comment]
Ah interesting! Thanks for the explanation :)

**Pull request merged**
